SJRK-288 & SJRK-359: added block reordering

BlueSlug opened the pull request:
This work addresses:
- [SJRK-288](https://issues.fluidproject.org/browse/SJRK-288) (block reordering with buttons)
- [SJRK-359](https://issues.fluidproject.org/browse/SJRK-359) (block reordering with drag & drop)
- [SJRK-365](https://issues.fluidproject.org/browse/SJRK-365) (focus on whole block groups with forward and backward keyboard navigation)

BlueSlug committed:
SJRK-359: added reorderer component to storyEditor, drag & drop working.
Still needs a bit of work

BlueSlug committed:
SJRK-288/SJRK-359: WIP added temp buttons, got them working, still buggy
- Cannot focus on the block input fields with mouse (no grab handle yet)
- There is a race condition that triggers an error when blocks are added 
too quickly
- The same error is fired when a story with more than one block is 
loaded from autosave
- The button styling and assets are not final

BlueSlug committed:
SJRK-288: Workaround for reorderer refresh on new block added.

Co-Authored-By: Justin Obara <obara.justin@gmail.com>

BlueSlug committed:
Merge remote-tracking branch 'upstream/master' into SJRK-288

BlueSlug committed:
SJRK-288: added tests for new blockUi.editor events and click handlers

BlueSlug committed:
SJRK-288/SJRK-359: added tests for block reorderer

BlueSlug committed:
SJRK-288: docs update after adding the storyEditor's reorderer component

BlueSlug committed:
SJRK-288: added avatar and dropMarker styling for block reorderer

cindyli commented on src/ui/ui-storyEditor.js:
```diff
@@ -153,6 +153,19 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
             blockManager: {
                 container: "{storyEditor}.options.selectors.storyEditorContent",
                 options: {
+                    distributeOptions: {
```
I wonder what's the reason to use `distributeOptions` to distribute listeners for `onMoveBlockDown` and `onMoveBlockUp` events rather than directly defining them on `sjrk.storyTelling.blockUi.editor` component? Is it because you only want to attach these listeners to the story editor, not the story viewer? But is there a chance that up/down buttons are still available and clickable on the story viewer?

I feel it's less confusing to define listeners along with corresponding events if possible.

amb26 replied:
See above suggestion - I think this linkage should be implemented by either forwarding one event firing to another or declaring one event as an alias to another (latter probably not recommended)

BlueSlug replied:
The reason `distributeOptions` is used here is twofold:
- the `blockUi.editor` components are dynamically created, and I wasn't able to craft an IoC reference in the `reorderer` component that would resolve properly
- I didn't want to add any references to the `reorderer` component in the `blockUi.editor` grade, as the `reorderer` is specific to `storyEditor`

I think @amb26's suggestion of forwarding the event through to the `storyEditor` may work best. I'll see what I can come up with that also allows me to combine the two events (as suggested in the review for `blockUi-editor.js`)

BlueSlug replied:
To answer @cindyli's question about whether we anticipate reordering will be used in the `storyViewer`: we do not

cindyli commented on tests/ui/js/ui-storyEditorTests.js:
```diff
@@ -146,30 +153,113 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                },
+                {
+                    // Call reorderBlockUp, listen for reorderer onMove
+                    func: "{storyEditor}.reorderer.reorderBlockUp",
```
Is it necessary to call `reorderer.reorderBlockUp()` in the test? Shall it already be handled by the `onMoveBlockDown` listener when the down button is clicked at line 224?

I think this test sequence is not clear as it also manually triggers a couple of internal actions such as calling this `reorderBlockUp()` function and `reorderer.container.focusin` later. 

This test sequence can be simplified to: add blocks -> click "down" button -> the reorderer is refreshed -> the block is moved down to the proper position -> click "up" button -> the reorderer is refreshed -> up to the proper position.

In the sequence, check the inactive button states if needed.

BlueSlug replied:
The call to the `refresh` invoker was to force a refresh after blocks are added and to ensure the invoker was operating properly. I agree that, as an integration test, it's redundant. What I propose instead is to move the jQueryTrigger of `focusin` to right after the blocks are added, that way the sequence more accurately simulates the user's interaction with the editor.

Because the blocks are added dynamically, we can't get away without refreshing the reorderer before clicking on the "down" button.

cindyli commented on docs/GRADES-UI.md:
```diff
@@ -50,7 +51,9 @@
   and keywords fields to their respective model values in the `story`. It makes use of a `dynamicViewComponentManager`
-  called **blockManager** to add blocks of varying types on demand.
+  called **blockManager** to add blocks of varying types on demand. The `reorderer` component is for reordering individual
+  blocks via buttons on each (see `sjrk.storyTelling.block.editor` above), using CTRL+UP or CTRL+DOWN or by dragging and
```
It would be helpful to have a text on the interface to explain how to move blocks using keyboard. On Mac, the key combinations working for me are CTRL+SHIFT+UP or CTRL+SHIFT+DOWN.

cindyli commented:
The inactive button states for the "up" button on the first block and the "down" button on the last block are not implemented.

amb26 commented on src/ui/ui-storyEditor.js:
```diff
@@ -224,6 +237,49 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
                 }
+            },
+            reorderer: {
+                type: "fluid.reorderList",
```
Lovely and hilarious to see this back after a decade :)
:cake:

amb26 commented on src/ui/blockUi-editor.js:
```diff
@@ -17,8 +17,35 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
         },
+        events: {
+            onReadyToBind: null,
+            onBlockControlsBound: null,
```
Given this event is apparently only of use in test cases I suggest that its definition and binding be moved into a grade defined in the test cases themselves.

BlueSlug replied:
Given that it isn't even necessary for the test cases, I have decided to remove it altogether. I think it was leftover from a previous factoring.

amb26 commented on src/ui/ui-storyEditor.js:
```diff
@@ -224,6 +237,49 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                    listeners: {
+                        // We should be refreshing the reorderer on each new block
+                        // being added, but we can't reliably know when a block
+                        // has finished being rendered. It would also be possible
```
Why not - do we have any asynchronous rendering operations? If so, how do they work? Seems likely we should be able to tag the refresh operation onto them rather than apply this potentially expensive operation on every focus.

BlueSlug replied:
@jobara and I were exploring this, we even attempted to incorporate a `mutationObserver` into the `storyEditor` before that became too unwieldy, but we believe the issue boils down to [`gpii.handlebars.renderer`'s use](https://github.com/GPII/gpii-handlebars/blob/master/src/js/client/renderer.js#L15-L23) of jQuery functions to affect markup changes. In this case, we're using [`$.html()`](https://api.jquery.com/html/).

Since jQuery doesn't afford any way to know that it's done its thing, we can't reliably know when a newly-added blockUi is ready to be fiddled with, and therefore when it's safe to refresh the reorderer (this becomes an issue once a `grabHandle` is specified for a `movable`, something to do with the reorderer's `labeller` component, as I recall).

I've messaged @the-t-in-rtf to discuss how we might solve this issue, but I hadn't provided him with much detail and our workaround made it a less pressing matter.

amb26 replied:
I can't quite follow your explanation - $.html() is synchronous so it has "done it's thing" immediately?

BlueSlug replied:
I believe I misunderstood the root cause of the issue. In the case that we call the `refresh` invoker after adding a new block, the problem would arise when we attempt to add many blocks very quickly, either on loading an autosave or simply clicking the add button very rapidly. The reorderer will attempt to refresh itself, and upon coming into contact with a block that has a container added but not actually rendered, we'll get the error.

@jobara and I have discussed implementing a kind of event queue to manage adding blocks one at a time and waiting until the queue is empty to call refresh

cindyli replied:
Bump. Please update the code comment with the root cause, a TODO and the proper fix you suggested.

amb26 commented on src/ui/blockUi-editor.js:
```diff
@@ -17,8 +17,35 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+        events: {
+            onReadyToBind: null,
+            onBlockControlsBound: null,
+            onMoveBlockDown: null,
```
I think a separate event for each direction is a little too bulky - we can bind onto a single event and supply it with the appropriate fluid.direction.* argument. This would eliminate the two invokers also - and finally we can just implement the actual "reorderBlock" activity as a listener, eliminating the last invoker.

BlueSlug replied:
This is an excellent suggestion, which I've implemented for the most part.

I'm hesitant to remove the final `reorderBlock` invoker, since leaving it in means there's another spot that a programmer can reach in to change things without having to know the details of `sjrk.storyTelling.ui.storyEditor.reorderBlock`'s arguments. This really only saves having to pass one more argument (you'd need a reference to the `reorderer` in order to call the invoker anyway), so it won't be a massive difference, which is why I'm on the fence.

cindyli commented:
When using mouse to click on the heading or the text input fields of a block, the focus keeps jumping to focus on the entire block, probably the reorderer drag-drop handle. To reproduce:

1. Go to the story editor and add a block in any type;
2. Use mouse to click focus on the heading input field;
3. The focus jumps to the entire block.

It has no problem to use keyboard to tab focus on input fields.

This issue is reproducible in latest firefox, chrome and safari.

BlueSlug committed:
SJRK-288: addressed a number of review items:
- consolidated moveBlock events into a single event with direction 
passed as an argument
- removed the unused onBlockControlsBound
- added macOS keybindings to the reorderer documentation

BlueSlug committed:
SJRK-359: added drag & drop grab handle and new "move" icon

BlueSlug committed:
NOJIRA: updated README.md to pin version of couchdb docker container

BlueSlug committed:
SJRK-288: fixed issues with block reorderer grabHandle

BlueSlug committed:
SJRK-288: added block order model values and functions to update them

BlueSlug committed:
SJRK-288: alleviated race condition in storyEditor tests, need to fix it

BlueSlug committed:
SJRK-288: initial work to disable reorder buttons if block is top/bottom

BlueSlug committed:
SJRK-288: block move up/down buttons are now set once template is ready

BlueSlug committed:
NOJIRA: tidied up binder onUiReadyToBind listeners for clarity

amb26 commented on src/ui/blockUi-editor.js:
```diff
@@ -16,9 +16,63 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+            "moveBlockDownDisabled": {
+                func: "{that}.setButtonDisabled",
+                args: ["{that}.dom.moveBlockDownButton", "{change}.value"],
+                excludeSource: ["init"],
```
Should be able to avoid partnering this with the "onReady" duplicate listeners, especially given you are on the FLUID-6145-era Infusion - what is the result if you simply remove the excludeSource annotation here?

BlueSlug replied:
I believe the difference is invisible to the user. The model values will be set early on when the DOM isn't ready (the buttons aren't rendered yet), so nothing happens when `.prop()` is called. The listeners hanging off of `onReadyToBind` are to set the buttons once the markup is rendered, and the modelListeners will handle any subsequent changes. I'm ambivalent about excluding init, at this point, though I suppose we might have some barely-perceptible performance gains by not handling it on init.

Can you please elaborate on what FLUID-6145 is allowing me to do or not do?

amb26 replied:
One of the main improvements in the FLUID-6148/FLUID-6145 revision ensures that the timing of the "init" model transaction is fully predictable with respect to a component's "onCreate" - that is, it always comes before it. This should mean that the function of your duplicate listeners in "onReadyToBind" on lines 65-74 should be handlable by the model listener here as part of "init" - especially if you can move whatever it is that the component needs to be "ready" into the component's resources - the only such thing should be a template loading, as far as I can see, since everything following that should be synchronous

BlueSlug replied:
We discussed this issue a bit [in IRC](http://irc-logs.fluidproject.org/%23fluid-work/%23fluid-work.2020-06-18.log) today, and came to the decision to defer this work for another Jira. I've filed [SJRK-364](https://issues.fluidproject.org/browse/SJRK-364) to capture your suggestions, please feel free to amend it if it isn't accurate :)

I'll be adding a comment in this part of the code linking to both this review item and the new Jira.

amb26 commented on src/ui/ui.js:
```diff
@@ -205,4 +207,74 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+     */
+    sjrk.storyTelling.ui.updateBlockOrder = function (that, completionEvent) {
+        var blockUis = that.reorderer.container.children();
+        var loopCounter = 0;
```
This odd way of structuring the iteration will likely throw off a reader - why not write a conventional "for" loop iterating over blockUIs, or else write blockUIs.forEach(function (blockUi, blockIndex) { .... ?

amb26 commented on src/ui/ui.js:
```diff
@@ -205,4 +207,74 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+        var managedClassNamePattern = sjrk.storyTelling.ui.getManagedClassNamePattern(that.blockManager);
+
+        fluid.each(blockUis, function (blockUi) {
+            var managedClassName = fluid.find(blockUi.classList, function (value) {
```
I haven't traced through all the details of this, but is the predicate here attempting to have the effect of https://api.jquery.com/hasClass/ ?

BlueSlug replied:
What I attempted to do is to get a unique class name from the collection that matched a certain pattern, and then feed that class name into the `blockManager`'s component registry to get a reference to the `blockUi.editor` component that matches that class name, in order to reach its `block` subcomponent. I effectively wrote startsWith, though I think that's not supported in IE. Whether we're still supporting IE is something I'm unsure of.

I don't think `hasClass()` does what I want (I could be wrong), but if you know of any other approaches that fit the bill I would be glad to learn!

jobara replied:
Infusion is no longer explicitly supporting IE. We haven't removed code for it, but will no longer be testing in IE or actively writing code to support it. That means that any new release (pre-release or stable) cannot be relied on to support IE. So it may be time to revisit and/or define the browser support for Story Telling.

jobara replied:
@BlueSlug if I'm following the logic of the function, what you want to do is to take the DOM position of an element, find what Infusion component corresponds to that element, and then update the model values of that component with the position index. So maybe the first question is to @amb26 about whether there is an easy way to find a component from its container.

Another question is about the approach as a whole. That is you seem to only need this to update the disabled/enabled state of the up/down arrows for reordering. Here are some thoughts to get your started, none of them may be good though: 
- You could continue with the DOM based approach and on reorder have the components inspect where they are in the parents list of children or maybe if they have a previous or next sibling. In general modelizing state is better. 
- Should a reorderable element care about it's position? Should a single block UI really have to know that it is reorderable and where it is in the list? Perhaps this reordering behaviour could be the responsibility of a parent component that sort of decorates its children with the ability to be reordered. And there by managing the state of the individual reorderable elements.

amb26 replied:
Code to maintain an index of viewComponent containers to their components can be adapted from the "viewMapper" in fluidViewDebugging.js here - https://github.com/fluid-project/infusion/blob/master/src/framework/core/js/FluidViewDebugging.js#L576
You can simplify this code considerably since there is extra material here to account for the possibility that the view component lies in an iframe.
However, I agree with @jobara 's comment that responsibility for all list-based functionality should be centralised in a parent component.

colinbdclark replied:
> Infusion is no longer explicitly supporting IE. We haven't removed code for it, but will no longer be testing in IE or actively writing code to support it. That means that any new release (pre-release or stable) cannot be relied on to support IE. So it may be time to revisit and/or define the browser support for Story Telling.

I see no reason to support IE in the Storytelling Tool at this point. We have so much to do, so anything we can do to simplify development seems like a good idea. Let's not spend time trying to support Internet Explorer.

BlueSlug committed:
SJRK-288: cross-browser CSS fixes, consistent buttons and focus-styles

BlueSlug committed:
SJRK-288: added 6 contrast theme versions of block control buttons

BlueSlug committed:
SJRK-288: CSS focus & contrast fixes, also added ARIA labels to controls

BlueSlug committed:
SJRK-288: converted `fluid.each` to a `for` loop for better readability

BlueSlug committed:
SJRK-288: added more tests for `blockUi.editor` grade

BlueSlug committed:
SJRK-288: Updated How-to story with instructions for block reordering

BlueSlug committed:
SJRK-288: split ui grade tests off from storyUi grade tests

jobara committed:
SJRK-288: Make all the reorderable blocks focusable.

This is to improve tab navigation and reordering when moving from adding 
content in block to reordering the blocks.

BlueSlug committed:
Merge pull request #32 from jobara/SJRK-288

SJRK-365: Make all the reorderable blocks focusable.

BlueSlug committed:
Merge branch 'SJRK-288' of https://github.com/BlueSlug/sjrk-story-telling into SJRK-288

BlueSlug committed:
SJRK-288: refactored ui onReadyToBind, added new ui and storyUi tests

BlueSlug committed:
SJRK-288: relocated reordering functions to storyEditor,

BlueSlug committed:
SJRK-288: added contrast theme rules for focus state

BlueSlug committed:
SJRK-288: added reorderer function unit tests

BlueSlug commented:
@jobara's fix in 3038c83 solves SJRK-365

BlueSlug committed:
SJRK-288: added manual tests for Story Editor, including reorderer tests

BlueSlug committed:
SJRK-288: added link to how-to story on AIHEC theme masthead

cindyli commented on src/ui/ui.js:
```diff
@@ -18,8 +18,13 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
         events: {
-            onReadyToBind: null,
-            onControlsBound: null
+            onTemplateRendered: null,
```
I wonder what's the point to introduce the new `onTemplateRendered` event if it's an equivalent of `onReadyToBind`?

BlueSlug replied:
In this case, the `onReadyToBind` event is being set up as a compound event, to which other sub-events can be added in implementing grades. I needed to add the definition of `onTemplateRendered` at this level in order to refer to it within `onReadyToBind`.

cindyli commented on src/ui/ui.js:
```diff
@@ -116,8 +61,17 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
-            onStoryUpdatedFromBlocks: null
+            onStoryUpdatedFromBlocks: null,
+            onBlockManagerCreated: null,
+            onReadyToBind: {
```
Following the comment above, is your intention of using `onReadyToBind` event is to use it as an aggregate event for `onBlockManagerCreated` + `onTemplateRendered`? But as `onTemplateRendered` is the CreateOnEvent for instantiating `blockManager`, `onBlockManagerCreated` event should always be fired after `onTemplateRendered`. It doesn't seem making sense to include `onTemplateRendered` in the aggregation.

BlueSlug replied:
Yes, the intent is to use it as an aggregate event. I also noticed that `onBlockManagerCreated` would always fire after `onTemplateRendered` for `ui` grades with a `blockManager` component (instances of `storyUi`), but the intent is to keep `onReadyToBind` from firing the moment the template is done rendering (which was previously the case in `ui`), as that is too early for the `storyEditor` grade.

I thought this solution was a wiser choice than overriding the listener in the `templateManager` component of `ui`, and it allows for a bit more flexibility in configuring the timing. It might also be premature optimization :) Would you like me to revert to the previous setup and find another way to delay `onReadyToBind` for the `storyEditor`?

cindyli replied:
It don't seem like a good idea to share the same `onReadyToBind` between `ui` and `storyUI` if one is dependent on the other and they are fired in sequence. I would suggest to use different events, such as `onUIReady` for `ui` component and `onStoryUIReady` for `storyUI` component.

BlueSlug replied:
Okay, I'll separate the usage of them. I may still leave the template render event name as `onTemplateRendered`, since that's more accurate, and limit `onReadyToBind` usage to `storyUi` grades

cindyli replied:
Feel free to use names that make sense. I've never been good at coming up with names.

cindyli commented on src/ui/ui.js:
```diff
@@ -162,6 +117,7 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
                         }
                     },
                     listeners: {
+                        "onCreate.escalate": "{ui}.events.onBlockManagerCreated",
```
Shall `onCreate.escalate` be triggered after `onCreate.createBlocksFromData`? Depending on the answer to the question above, if `storyUi.events.onReadyToBind` doesn't need to be an aggregate event combined with `onTemplateRendered`, this escalation can directly goes to `{ui}.events.onReadyToBind` and remove the intervening `onBlockManagerCreated` event.

cindyli commented on src/ui/ui.js:
```diff
@@ -205,4 +161,64 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+     */
+    sjrk.storyTelling.ui.storyUi.getBlockGradeFromEventModelValues = function (modelValuesFromEvent) {
+        var gradeName = "sjrk.storyTelling.block-" + fluid.allocateGuid();
+        fluid.defaults(gradeName, {
```
I'm a bit surprised that this grade declaration doesn't require a `gradeNames` such as:
```
fluid.defaults(gradeName, {
    gradeNames: ["fluid.modelComponent"],
    ...
```

BlueSlug replied:
Me too! The result of this function is passed into a `gradeNames` key in the `block` subcomponents of the `blockManager`, for what it's worth. The exact details require some careful parsing of the `blockManager` component and the grade it's based on, `sjrk.dynamicViewComponentManager`.

It is a weird setup, but it's the one we figured out that lets us dynamically add viewComponents with varying grades.

amb26 replied:
This isn't recommended since in current Infusion these defaults definitions will leak - whilst dynamically generated grades are used from time to time, this is only where there is no alternative. It seems like a better approach would be to ensure that the values fired in the event are already properly defaulted (in createBlocksFromData ?) and then to write the dynamic references to the event arguments inline in the definition of the "block" component.
General notes - 
- yes, supplying no grades at all in "mixin" grades is a reasonable technique, and was even sometimes required in the old framework where it was possible that trying to mix in "fluid.modelComponent" to a view component again would trash its argument list. This bug got fixed in the new framework.
- the definition of "storyUi" is too large and should be broken up at least at one point, perhaps at "blockManager", "managedViewComponents", or both. It would be helpful to supply at least some kind of base grade for "managedViewComponents" in any case to guide the reader as to what kind of material might get created.

BlueSlug replied:
I agree with these suggestions (and thank you for them!), though I feel they're out of scope for this Jira.

I went to file a Jira to capture your feedback, but there are in fact four(!) that already address reviewing the DVCM and its usage as the blockManager, and hopefully the time to tackle them will come soon :)

The most relevant issues are [SJRK-115](https://issues.fluidproject.org/browse/SJRK-115) and [SJRK-126](https://issues.fluidproject.org/browse/SJRK-126), while [SJRK-61](https://issues.fluidproject.org/browse/SJRK-61) and [SJRK-262](https://issues.fluidproject.org/browse/SJRK-262) are a little more tangentially relevant.

jobara replied:
@BlueSlug if these issues will not be addressed in this work, it would be useful to track the related issues in comments with the relevant code.

cindyli commented on src/ui/ui.js:
```diff
@@ -205,4 +161,64 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+    sjrk.storyTelling.ui.storyUi.getBlockGradeFromEventModelValues = function (modelValuesFromEvent) {
+        var gradeName = "sjrk.storyTelling.block-" + fluid.allocateGuid();
+        fluid.defaults(gradeName, {
+            // TODO: this should test that modelValuesFromEvent is a legitimate
```
This comment sentence is incomplete.

BlueSlug replied:
The rest of the comment is on the next line. Should I restyle it somehow or combine it into one line? This should probably be a new Jira, actually

cindyli replied:
Leave as is as long as it makes sense...

cindyli commented on src/ui/ui-storyEditor.js:
```diff
@@ -68,14 +68,7 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
-                    onBlockManagerCreated: "{that}.events.onBlockManagerCreated"
-                }
-            }
+            onBlockOrderUpdated: null
```
It seems `updateBlockOrder` is only used in the test. I wonder instead of listening to this event in the test, would you be able to test if the story blocks have been updated/sorted correctly?

BlueSlug replied:
I began to work on this and added some explicit test cases that check the story contents before and after each reorder, but in doing so I surfaced a previously-unknown bug: it seems that there is something happening with the changeApplier call in `sjrk.storyTelling.ui.storyEditor.sortStoryContent` that is merging the contents of that value rather than replacing it. The result is that blocks in the `story`'s content array have nearly every possible key present, i.e. text blocks have their usual fields in addition to `imageUrl` **and** `mediaUrl`, `fileDetails`, `alternativeText`, etc., and this ends up in the database on publish.

I have tried, among other things:
- calling the changeApplier with "DELETE" on the `content` key before setting the new value
- bundling the individual block component order value updates as a single change transaction
- using `fluid.set` to set the `content` model value explicitly
- specifying the story grade's `mergePolicy` as `nomerge` (maybe this should have been `replace`...)

The consequences are minimal, basically only that the test story will look strange since certain block types will have unexpected fields, though we should find a solution before it causes any real problems.

amb26 replied:
The fact that it is "accumulating" is a well-known design feature of the ChangeApplier since its earliest days - whether it is the best choice or not is secondary to the issue that doing anything about it at this point would break thousands of lines of code depending on it.
The correct approach is a combination of your first and second points - as seen in utilities such as this one in the prefs framework store - https://github.com/fluid-project/infusion/blob/master/src/framework/preferences/js/Store.js#L245 - this is such a frequently used pattern that it should be a core framework utility.
Point 3, using fluid.set directly into the model must never be operated since it will corrupt the ChangeApplier's records, and point 4 is inadvisable since mergePolicies are generally framework internals that should only be used as a last resort - their operation is fragile and not terribly well documented.
I agree we should get rid of this event.

BlueSlug replied:
The solution turned out to be to add a `DELETE` operation before updating the story contents in `sjrk.storyTelling.ui.storyUi.updateStoryFromBlocks`, not in `sjrk.storyTelling.ui.storyEditor.sortStoryContent`. I have wrapped up all changes in transactions either way, in order to economise model changes and avoid any unnecessary firing of the block-to-story model listener.

cindyli commented on tests/ui/js/ui-storyEditorTests.js:
```diff
@@ -146,30 +153,103 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                },
+                {
+                    // Refresh the reorderer, listen for onBlockOrderUpdated after it's done
+                    func: "{storyEditor}.reorderer.refresh"
```
Do you need to manually trigger the `reorderer.refresh()`? Shouldn't it be triggered automatically when items are reordered?

cindyli commented on tests/ui/js/ui-storyEditorTests.js:
```diff
@@ -225,6 +305,92 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+                expectedContent: [{ id: "a", order: 0 }, { id: "b", order: 0 }]
+            },
+            bothFalse: {
+                content: [{ id: "a", order: false }, { id: "b", order: false }],
```
what if one is `true` and the other is `false`? Thanks for writing such a comprehensive coverage. It reminds me that the sorting probably should return without proceeding if the `order` value is not a valid number. If a non numerical value reaches this function, something goes very wrong. ;)

cindyli commented:
The focus styling for "move up/down" buttons are quite vague and hard to notice. The focus styling for other buttons: add blocks, delete and update, has the same issue. If I remember correct, missing hover styling for all buttons is a known issue that has a separate ticket.

cindyli commented:
The focus styling for "move up/down" buttons on UIO THEMES are totally unnoticeable.

cindyli commented:
When dragging the mouse to select some buttons, the "play" button from the text-to-speech shows up. Clicking on it doesn't read anything. If this is a problem, it probably should be addressed in a separate JIRA.

<img width="340" alt="Screen Shot 2020-06-23 at 9 20 53 PM" src="https://user-images.githubusercontent.com/572877/85486044-e121cf80-b597-11ea-9432-7d1e91da22a3.png">

cindyli commented:
Refreshing the story editor replicates existing blocks:

1. Go to the story editor and add one block;
2. Enter some text in the block;
3. Refresh the page using CTRL+r or browser refresh button;
4. Now there are 2 blocks in the editor. The 2nd one is a replica of the first one with the same text.

Expected: There should always be same blocks in the editor before and after the refresh.

BlueSlug committed:
NOJIRA: updated BBR and GD contrast theme buttons to fix visual bug

BlueSlug committed:
NOJIRA: Solved a CSS bug - forward/back arrow misalignment

BlueSlug committed:
SJRK-288: solved a CSS bug - Firefox button focus states were invisible

BlueSlug committed:
SJRK-288: separated ready events in ui and storyUi grades

BlueSlug committed:
SJRK-288: fixed story double-load bug introduced by event refactoring

BlueSlug committed:
SJRK-288: updated reorderer race condition comment with a link to Jira

BlueSlug commented:
> When dragging the mouse to select some buttons, the "play" button from the text-to-speech shows up. Clicking on it doesn't read anything. If this is a problem, it probably should be addressed in a separate JIRA.
> 
> <img alt="Screen Shot 2020-06-23 at 9 20 53 PM" width="340" src="https://user-images.githubusercontent.com/572877/85486044-e121cf80-b597-11ea-9432-7d1e91da22a3.png">

We still haven't patched in the fix for FLUID-6512, which would eliminate the page reload when `Play` is clicked, but I'm not sure it's a problem that an empty selection isn't readying anything. The Orator is not designed to read anything other than what is displayed on the screen, if I recally correctly, so no alt-text or aria things. @jobara may be able to confirm this behaviour.

jobara commented:
> > When dragging the mouse to select some buttons, the "play" button from the text-to-speech shows up. Clicking on it doesn't read anything. If this is a problem, it probably should be addressed in a separate JIRA.
> > <img alt="Screen Shot 2020-06-23 at 9 20 53 PM" width="340" src="https://user-images.githubusercontent.com/572877/85486044-e121cf80-b597-11ea-9432-7d1e91da22a3.png">
> 
> We still haven't patched in the fix for FLUID-6512, which would eliminate the page reload when `Play` is clicked, but I'm not sure it's a problem that an empty selection isn't readying anything. The Orator is not designed to read anything other than what is displayed on the screen, if I recally correctly, so no alt-text or aria things. @jobara may be able to confirm this behaviour.

The Orator only reads the visible text. I suppose you should be able to configure that behaviour although it would likely require some new functions be implemented to know how to parse the text. 

The play button shows up with a selection, if you are finding that it shows up even when no text is selected, perhaps a JIRA against the Orator is warranted.

BlueSlug committed:
SJRK-288: removed unused events, batched all order-related changes

BlueSlug committed:
SJRK-288: added basic input-checking in the `sortStoryContent` function

BlueSlug committed:
SJRK-288: added regression test for double-loading blockUi bug

BlueSlug committed:
SJRK-288: added persistent highlight for Help story link in AIHEC theme

BlueSlug commented:
Ready for review

cindyli commented on src/ui/ui-storyEditor.js:
```diff
@@ -252,4 +293,107 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+
+            var block = that.blockManager.managedViewComponentRegistry[managedClassName].block;
+
+            var firstInOrder = i === 0 ? true : false;
```
Line 332-341 can be simplified to:

```
block.applier.change({
    path: "",
    value: {
        order: i,
        firstInOrder: i === 0 ? true : false;
        lastInOrder: i === blockUis.length - 1 ? true : false;
    }
});
```

BlueSlug replied:
This doesn't seem to work: Editor tests are failing and it breaks the reorderer buttons. I'll continue to investigate.

cindyli replied:
This form works: 

```
block.applier.change("", {
  order: i,
  firstInOrder: i === 0 ? true : false,
  lastInOrder: i === blockUis.length - 1 ? true : false
});
```

amb26 replied:
@cindyli 's code should be identical in all observable effects - I'd be interested in reason for the discrepancy once you find it.
Note that the expression can be simplified yet further:
````
block.applier.change("", 
        order: i,
        firstInOrder: i === 0,
        lastInOrder: i === blockUis.length - 1
});
````
(@cindyli 's version had a few stray semicolons but I imagine you removed them)
(Edited further to compensate for effect of the other typo where "path", "value" where framed at top level)

BlueSlug replied:
Removing the `path` and `value` references did the trick. Thanks @cindyli and @amb26 :)

BlueSlug replied:
> > > When dragging the mouse to select some buttons, the "play" button from the text-to-speech shows up. Clicking on it doesn't read anything. If this is a problem, it probably should be addressed in a separate JIRA.
> > > <img alt="Screen Shot 2020-06-23 at 9 20 53 PM" width="340" src="https://user-images.githubusercontent.com/572877/85486044-e121cf80-b597-11ea-9432-7d1e91da22a3.png">
> > 
> > 
> > We still haven't patched in the fix for FLUID-6512, which would eliminate the page reload when `Play` is clicked, but I'm not sure it's a problem that an empty selection isn't readying anything. The Orator is not designed to read anything other than what is displayed on the screen, if I recally correctly, so no alt-text or aria things. @jobara may be able to confirm this behaviour.
> 
> The Orator only reads the visible text. I suppose you should be able to configure that behaviour although it would likely require some new functions be implemented to know how to parse the text.
> 
> The play button shows up with a selection, if you are finding that it shows up even when no text is selected, perhaps a JIRA against the Orator is warranted.

This has been captured in [FLUID-6523](https://issues.fluidproject.org/browse/FLUID-6523)

cindyli commented on tests/ui/html/ui-storyEditor-manual-Tests.html:
```diff
@@ -0,0 +1,202 @@
+<html>
```
Could you add a documentation to describe available manual tests, their goals and how to access, etc? Thanks.

cindyli commented:
The insertion line when reordering blocks disappears when a UIO theme is selected:

1. Go to the story editor and insert a couple of blocks;
2. Open UIO and select a theme;
3. Drag and drop a block to reorder it;
4. The insertion line that shows up to indicate the drop location doesn't show up.

amb26 commented on src/ui/ui-storyEditor.js:
```diff
@@ -252,4 +293,107 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+
+        // used to retrieve the class name for each block
+        // within the blockManager's managedViewComponentRegistry
+        var managedClassNamePattern = sjrk.storyTelling.ui.storyEditor.getManagedClassNamePattern(that.blockManager.options.selectors.managedViewComponents);
```
This code has changed a lot since the first review but I still think the "managedClassNamePattern" strategy is quite overcomplicated compared with the requirement - did the alternative strategies, e.g. via a "viewMapper" component or similar get written up as a JIRA somewhere?

BlueSlug replied:
It has now! I've just filed [SJRK-371](https://issues.fluidproject.org/browse/SJRK-371) and have left a comment in the code pointing to it. Many thanks again for your feedback, hopefully we can get to that soon.

BlueSlug committed:
SJRK-288: added UIO themes for block reorder drop marker

BlueSlug committed:
SJRK-288: Tidied updateBlockOrder code, added manual test documentation

BlueSlug committed:
SJRK-288: added TODOs and Jira links to outstanding issues

BlueSlug commented:
> > > When dragging the mouse to select some buttons, the "play" button from the text-to-speech shows up. Clicking on it doesn't read anything. If this is a problem, it probably should be addressed in a separate JIRA.
> > > <img alt="Screen Shot 2020-06-23 at 9 20 53 PM" width="340" src="https://user-images.githubusercontent.com/572877/85486044-e121cf80-b597-11ea-9432-7d1e91da22a3.png">
> > 
> > 
> > We still haven't patched in the fix for FLUID-6512, which would eliminate the page reload when `Play` is clicked, but I'm not sure it's a problem that an empty selection isn't readying anything. The Orator is not designed to read anything other than what is displayed on the screen, if I recally correctly, so no alt-text or aria things. @jobara may be able to confirm this behaviour.
> 
> The Orator only reads the visible text. I suppose you should be able to configure that behaviour although it would likely require some new functions be implemented to know how to parse the text.
> 
> The play button shows up with a selection, if you are finding that it shows up even when no text is selected, perhaps a JIRA against the Orator is warranted.

This has been captured in [FLUID-6523](https://issues.fluidproject.org/browse/FLUID-6523)

BlueSlug commented:
Ready for review
